govcd/nsxv_ipset.go

```diff
@@ -0,0 +1,254 @@
+// and find the IP set by ID. If it is not found - then a search by name is performed.
+func (vdc *Vdc) GetNsxvIpSetByNameOrId(identifier string) (*types.EdgeIpSet, error) {
+	getByName := func(name string, refresh bool) (interface{}, error) { return vdc.GetNsxvIpSetByName(name) }
+	getById := func(id string, refresh bool) (interface{}, error) { return vdc.GetNsxvIpSetById(id) }
+	entity, err := getEntityByNameOrId(getByName, getById, identifier, true)
+	if entity == nil {
+		return nil, err
+	}
+	return entity.(*types.EdgeIpSet), err
+}
+
+// GetAllNsxvIpSets retrieves all IP sets and returns []*types.EdgeIpSet or an
+// error of type ErrorEntityNotFound if there are no IP sets
+func (vdc *Vdc) GetAllNsxvIpSets() ([]*types.EdgeIpSet, error) {
+	vdcId, err := getUuidFromHref(vdc.Vdc.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get vdc ID from HREF: %s", err)
+	}
+
+	// build a path for to read all IP sets in a scope. A scope is defined by vDC ID. The endpoint
+	// should look like:
+	// https://192.168.1.109/network/services/ipset/scope/f9daf2da-b4f9-4921-a2f4-d77a943a381c where
+	// the trailing UUID is vDC ID
+	httpPath, err := vdc.buildNsxvNetworkServiceEndpointURL(types.NsxvIpSetServicePath + "/scope/" + vdcId)
+	if err != nil {
+		return nil, fmt.Errorf("could not get network services API endpoint for IP set: %s", err)
+	}
+
+	// Anonymous struct to unwrap list of IP sets <list><ipset></ipset><ipset></ipset></list>
+	ipSetsResponse := &struct {
+		XMLName          xml.Name `xml:"list"`
+		types.EdgeIpSets `xml:"ipset"`
+	}{}
+
+	// This query returns all IP sets on the scope (scoped by vDC ID)
+	errString := fmt.Sprintf("unable to read IP sets for scope %s: %%s", vdcId)
+	_, err = vdc.client.ExecuteRequest(httpPath, http.MethodGet, types.AnyXMLMime, errString, nil, ipSetsResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ipSetsResponse.EdgeIpSets) == 0 {
+		return nil, ErrorEntityNotFound
+	}
+
+	return ipSetsResponse.EdgeIpSets, nil
+}
+
+// DeleteNsxvIpSetById deletes IP set by its ID which is formatted as
+// f9daf2da-b4f9-4921-a2f4-d77a943a381c:ipset-9
+func (vdc *Vdc) DeleteNsxvIpSetById(id string) error {
+	err := validateDeleteNsxvIpSet(id, "")
+	if err != nil {
+		return err
+	}
+
+	// build a path for to delete exact IP set sample path is: DELETE API-URL/services/ipset/id:ipset-#
+	// https://192.168.1.109/network/services/ipset/f9daf2da-b4f9-4921-a2f4-d77a943a381c:ipset-9
+	httpPath, err := vdc.buildNsxvNetworkServiceEndpointURL(types.NsxvIpSetServicePath + "/" + id)
+	if err != nil {
+		return fmt.Errorf("could not get network services API endpoint for IP set: %s", err)
+	}
+
+	errString := fmt.Sprintf("unable to delete IP set with ID %s: %%s", id)
+	_, err = vdc.client.ExecuteRequestWithCustomError(httpPath, http.MethodDelete, types.AnyXMLMime,
+		errString, nil, &types.NSXError{})
+	if err != nil {
+		return err
```

Comment:
Custom message on what action error happen would be helpful

Reply:
The error message is declare in line 196, then fed into `ExecuteRequestWithCustomError` and the error already comes back with a message.